File: samples/javascript_nodejs/43.complex-dialog/dialogs/reviewSelectionDialog.js
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

const { ChoicePrompt, ComponentDialog, WaterfallDialog } = require('botbuilder-dialogs');

const REVIEW_SELECTION_DIALOG = 'REVIEW_SELECTION_DIALOG';

const CHOICE_PROMPT = 'CHOICE_PROMPT';
const WATERFALL_DIALOG = 'WATERFALL_DIALOG';

class ReviewSelectionDialog extends ComponentDialog {
    constructor() {
        super(REVIEW_SELECTION_DIALOG);

        // Define a "done" response for the company selection prompt.
        this.doneOption = 'done';

        // Define value names for values tracked inside the dialogs.
        this.companiesSelected = 'value-companiesSelected';

        // Define the company choices for the company selection prompt.
        this.companyOptions = ['Adatum Corporation', 'Contoso Suites', 'Graphic Design Institute', 'Wide World Importers'];

        this.addDialog(new ChoicePrompt(CHOICE_PROMPT));
        this.addDialog(new WaterfallDialog(WATERFALL_DIALOG, [
            this.selectionStep.bind(this),
            this.loopStep.bind(this)
        ]));

        this.initialDialogId = WATERFALL_DIALOG;
    }

    async selectionStep(stepContext) {
        // Continue using the same selection list, if any, from the previous iteration of this dialog.
        const list = Array.isArray(stepContext.options) ? stepContext.options : [];
        stepContext.values[this.companiesSelected] = list;

        // Create a prompt message.
        let message = '';
        if (list.length === 0) {
            message = `Please choose a company to review, or \`${ this.doneOption }\` to finish.`;
        } else {
            message = `You have selected **${ list[0] }**. You can review an additional company, or choose \`${ this.doneOption }\` to finish.`;
        }

        // Create the list of options to choose from.
        const options = list.length > 0
            ? this.companyOptions.filter(function(item) { return item !== list[0]; })
            : this.companyOptions.slice();
        options.push(this.doneOption);

        // Prompt the user for a choice.
        return await stepContext.prompt(CHOICE_PROMPT, {
            prompt: message,
            retryPrompt: 'Please choose an option from the list.',
            choices: options
        });
    }

    async loopStep(stepContext) {
        // Retrieve their selection list, the choice they made, and whether they chose to finish.
        const list = stepContext.values[this.companiesSelected];
        const choice = stepContext.result;
        const done = choice.value === this.doneOption;

        if (!done) {
            // If they chose a company, add it to the list.
            list.push(choice.value);
        }

        if (done || list.length > 1) {
            // If they're done, exit and return their list.
            return await stepContext.endDialog(list);
        } else {
            // Otherwise, repeat this dialog, passing in the list from this iteration.
            return await stepContext.replaceDialog(REVIEW_SELECTION_DIALOG, list);
        }
    }
}

module.exports.ReviewSelectionDialog = ReviewSelectionDialog;
module.exports.REVIEW_SELECTION_DIALOG = REVIEW_SELECTION_DIALOG;

File: samples/java_springboot/43.complex-dialog/src/main/java/com/microsoft/bot/sample/complexdialog/ReviewSelectionDialog.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package com.microsoft.bot.sample.complexdialog;

import com.microsoft.bot.builder.MessageFactory;
import com.microsoft.bot.dialogs.ComponentDialog;
import com.microsoft.bot.dialogs.DialogTurnResult;
import com.microsoft.bot.dialogs.WaterfallDialog;
import com.microsoft.bot.dialogs.WaterfallStep;
import com.microsoft.bot.dialogs.WaterfallStepContext;
import com.microsoft.bot.dialogs.choices.ChoiceFactory;
import com.microsoft.bot.dialogs.choices.FoundChoice;
import com.microsoft.bot.dialogs.prompts.ChoicePrompt;
import com.microsoft.bot.dialogs.prompts.PromptOptions;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import org.apache.commons.lang3.StringUtils;

public class ReviewSelectionDialog extends ComponentDialog {
    // Define a "done" response for the company selection prompt.
    private static final String DONE_OPTION = "done";

    // Define value names for values tracked inside the dialogs.
    private static final String COMPANIES_SELECTED = "value-companiesSelected";

    private static List<String> companiesOptions = Arrays.asList(
        "Adatum Corporation", "Contoso Suites", "Graphic Design Institute", "Wide World Importers"
    );

    public ReviewSelectionDialog(String withId) {
        super(withId);

        addDialog(new ChoicePrompt("ChoicePrompt"));

        WaterfallStep[] waterfallSteps = {
            this::selectionStep,
            this::loopStep
        };
        addDialog(new WaterfallDialog("WaterfallDialog", Arrays.asList(waterfallSteps)));

        // The initial child Dialog to run.
        setInitialDialogId("WaterfallDialog");
    }

    private CompletableFuture<DialogTurnResult> selectionStep(WaterfallStepContext stepContext) {
        // Continue using the same selection list, if any, from the previous iteration of this dialog.
        List<String> list = stepContext.getOptions() instanceof List
            ? (List<String>) stepContext.getOptions()
            : new ArrayList<>();
        stepContext.getValues().put(COMPANIES_SELECTED, list);

        String message;
        if (list.size() == 0) {
            message = String.format("Please choose a company to review, or `%s` to finish.", DONE_OPTION);
        } else {
            message = String.format("You have selected **%s**. You can review an additional company, or choose `%s` to finish.",
                list.get(0),
                DONE_OPTION);
        }

        // Create the list of options to choose from.
        List<String> options = new ArrayList<>(companiesOptions);
        options.add(DONE_OPTION);
        if (list.size() > 0) {
            options.remove(list.get(0));
        }

        PromptOptions promptOptions = new PromptOptions();
        promptOptions.setPrompt(MessageFactory.text(message));
        promptOptions.setRetryPrompt(MessageFactory.text("Please choose an option from the list."));
        promptOptions.setChoices(ChoiceFactory.toChoices(options));

        // Prompt the user for a choice.
        return stepContext.prompt("ChoicePrompt", promptOptions);
    }

    private CompletableFuture<DialogTurnResult> loopStep(WaterfallStepContext stepContext) {
        // Retrieve their selection list, the choice they made, and whether they chose to finish.
        List<String> list = (List<String>) stepContext.getValues().get(COMPANIES_SELECTED);
        FoundChoice choice = (FoundChoice) stepContext.getResult();
        boolean done = StringUtils.equals(choice.getValue(), DONE_OPTION);

        // If they chose a company, add it to the list.
        if (!done) {
            list.add(choice.getValue());
        }

        // If they're done, exit and return their list.
        if (done || list.size() >= 2) {
            return stepContext.endDialog(list);
        }

        // Otherwise, repeat this dialog, passing in the list from this iteration.
        return stepContext.replaceDialog(getId(), list);
    }
}
